subprojects/tooling-api/src/main/java/org/gradle/tooling/ProjectConnection.java

```diff
@@ -155,6 +159,32 @@
      */
     BuildActionExecuter.Builder action();
 
+    /**
+     * Notifies all daemons about file changes made by an external process, like an IDE.
+     *
+     * <p>The daemons will use this information to update the retained file system state.
+     *
+     * <p>The paths which are passed in need to be absolute, canonicalized paths.
+     * For a delete, the deleted path should be passed.
+     * For a rename, the old and the new path should be passed.
+     * When creating a new file, the path to the file should be passed.
+     *
+     * <p>The call is synchronous, i.e. the method ensures that the changed paths are taken into account
+     * by the daemon after it returned. This ensures that for every build started
+     * after this method has been called knows about the changed paths.
+     *
+     * <p>If the version of Gradle does not support virtual file system retention (i.e. &lt; 6.1),
+     * then the operation is a no-op.
+     *
+     * @param changedPaths Absolute paths which have been changed by the external process.
+     * @throws IllegalArgumentException When the paths are not absolute.
```

Comment:
Does it also throw is path is not canonical?

Reply:
Currently, it doesn't.

Reply:
I think it shouldn't throw if paths are not canonical. We might need to canonicalize, though, to handle symlinks properly.